File: sys/src/nonlinear/isam2.cc
#include "isam2.h"

namespace gtsam
{

    std::unique_ptr<ISAM2> default_isam2()
    {
        return std::make_unique<ISAM2>();
    }

    void update_noresults(ISAM2 &isam2, const NonlinearFactorGraph &graph, const Values &initial_values)
    {
        ISAM2Result result = isam2.update(graph, initial_values);
    }

    std::unique_ptr<Values> calculate_estimate(const ISAM2 &isam2) {
        return std::make_unique<Values>(isam2.calculateEstimate());
    }
}

File: src/nonlinear/isam2.rs
use cxx::UniquePtr;

use crate::nonlinear::{values::{Values, ValuesRef}, nonlinear_factor_graph::NonlinearFactorGraph};

pub struct ISAM2 {
    pub(super) inner: UniquePtr<::sys::ISAM2>,
}

impl Default for ISAM2 {
    fn default() -> Self {
        Self {
            inner: ::sys::default_isam2(),
        }
    }
}

impl ISAM2 {
    pub fn update_noresults(
        &mut self,
        graph: &NonlinearFactorGraph,
        values: &Values,
    ) {
        ::sys::update_noresults(self.inner.pin_mut(), &graph.inner, &values.inner);
    }

    pub fn calculate_estimate(
        &self,
    ) -> Values {
        Values {
            inner: ::sys::calculate_estimate(& self.inner),
        }
    }
}
